quarkchain/cluster/prom.py

```diff
@@ -0,0 +1,141 @@
+import logging
+import argparse
+import time
+from quarkchain.utils import token_id_encode
+from typing import List, Tuple
+from prometheus_client import start_http_server, Gauge
+from quarkchain.tools.count_total_balance import get_jsonrpc_cli, count_total_balance
+
+import jsonrpcclient
+
+# disable jsonrpcclient verbose logging
+logging.getLogger("jsonrpcclient.client.request").setLevel(logging.WARNING)
+logging.getLogger("jsonrpcclient.client.response").setLevel(logging.WARNING)
+
+TIMEOUT = 10
+
+host = "http://localhost:38391"
+
+
+def get_latest_minor_block_id_from_root_block(
+    root_block_height: int,
+) -> Tuple[int, List[str]]:
+    global host
+    cli = get_jsonrpc_cli(host)
+    res = cli.send(
+        jsonrpcclient.Request("getRootBlockByHeight", hex(root_block_height)),
+        timeout=TIMEOUT,
+    )
+    if not res:
+        raise RuntimeError("Failed to query root block at height" % root_block_height)
+
+    # Chain ID + shard ID uniquely determines a shard.
+    shard_to_header = {}
+    for mh in res["minorBlockHeaders"]:
+        # Assumes minor blocks are sorted by shard and height.
+        shard_to_header[mh["chainId"] + mh["shardId"]] = mh["id"]
+
+    return res["timestamp"], list(shard_to_header.values())
+
+
+def get_time_and_balance(root_block_height, token_id):
+    # TODO: handle cases if the root block doesn't contain all the shards.
+    time_stamp, minor_block_ids = get_latest_minor_block_id_from_root_block(
+        root_block_height
+    )
+
+    total_balances = {}
+    for block_id in minor_block_ids:
+        shard = "0x" + block_id[-8:]
+        total, starter, cnt = 0, None, 0
+        while starter != "0x" + "0" * 40:
+            balance, starter = count_total_balance(block_id, token_id, starter)
+            # may add same gap here to slow down the query
+            total += balance
+            cnt += 1
+        # Can't suppress scientific notation in prometheus currently, using a suitable rate now
+        total_balances[shard] = total / 10 ** 23
```

Comment:
I thought division gives you float numbers. 1) we want integer; 2) why need the division

Reply:
My bad, now I found that although numbers bigger than 1M would be presented as scientific notation in 8000 port, the data stored in the Prometheus server can still be piped by the original format. So maybe we can let Grafana handling the scaling problem and discard division here?
![image](https://user-images.githubusercontent.com/66969090/88094850-14309200-cb49-11ea-99fc-7e39bf2d043b.png)